### ISSUE TITLE
trainer.predict() fails if DataLoader has a numpy array attribute

### ISSUE THREAD

**Issue by gokceneraslan** [Lightning-AI/lightning#15408]
### Bug description

I am using a custom DataLoader (https://github.com/scverse/scvi-tools/blob/main/scvi/dataloaders/_ann_dataloader.py#L89) with PL and I'm trying to run the `predict()` method of PL trainer. However it throws an exception and the reason is the mishandling of an array attribute in the DataLoader. See the reproducible code below. Ideally the predict() method should be robust enough for any custom dataloader.

### How to reproduce the bug

```python
import os
from typing import Any

import torch
from torch.utils.data import DataLoader

from pytorch_lightning import LightningModule, Trainer
import numpy as np


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)


class AnnLoader(DataLoader):
    def __init__(self, adatas: Any, batch_size, indices=None, shuffle=False, **kwargs):
        super().__init__(adatas, batch_size=batch_size, shuffle=shuffle, **kwargs)
        self.indices = np.random.rand(10, 10) if indices is None else indices


def run():
    data = AnnLoader([1, 2, 3], batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_predict_batches=1,
        replace_sampler_ddp=False
    )
    trainer.predict(model, data)


if __name__ == "__main__":
    run()
```

### Error messages and logs

```pytb
GPU available: True (cuda), used: False
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
`Trainer(limit_predict_batches=1)` was configured so 1 batch will be used.
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
/tmp/ipykernel_3102/784448938.py in <module>
     36 
     37 if __name__ == "__main__":
---> 38     run()

/tmp/ipykernel_3102/784448938.py in run()
     32         replace_sampler_ddp=False
     33     )
---> 34     trainer.predict(model, data)
     35 
     36 

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in predict(self, model, dataloaders, datamodule, return_predictions, ckpt_path)
    948         self.strategy.model = model or self.lightning_module
    949         return self._call_and_handle_interrupt(
--> 950             self._predict_impl, model, dataloaders, datamodule, return_predictions, ckpt_path
    951         )
    952 

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in _call_and_handle_interrupt(self, trainer_fn, *args, **kwargs)
    648                 return self.strategy.launcher.launch(trainer_fn, *args, trainer=self, **kwargs)
    649             else:
--> 650                 return trainer_fn(*args, **kwargs)
    651         # TODO(awaelchli): Unify both exceptions below, where `KeyboardError` doesn't re-raise
    652         except KeyboardInterrupt as exception:

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in _predict_impl(self, model, dataloaders, datamodule, return_predictions, ckpt_path)
    994         self._predicted_ckpt_path = self.ckpt_path  # TODO: remove in v1.8
    995 
--> 996         results = self._run(model, ckpt_path=self.ckpt_path)
    997 
    998         assert self.state.stopped

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in _run(self, model, ckpt_path)
   1164         self._checkpoint_connector.resume_end()
   1165 
-> 1166         results = self._run_stage()
   1167 
   1168         log.detail(f"{self.__class__.__name__}: trainer tearing down")

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in _run_stage(self)
   1249             return self._run_evaluate()
   1250         if self.predicting:
-> 1251             return self._run_predict()
   1252         return self._run_train()
   1253 

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in _run_predict(self)
   1305 
   1306     def _run_predict(self) -> Optional[_PREDICT_OUTPUT]:
-> 1307         self.reset_predict_dataloader(self.lightning_module)
   1308         # reset trainer on this loop and all child loops in case user connected a custom loop
   1309         self.predict_loop.trainer = self

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/trainer.py in reset_predict_dataloader(self, model)
   1960         if source.is_defined() and enable_prediction:
   1961             self.num_predict_batches, self.predict_dataloaders = self._data_connector._reset_eval_dataloader(
-> 1962                 RunningStage.PREDICTING, model=pl_module
   1963             )
   1964 

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/connectors/data_connector.py in _reset_eval_dataloader(self, mode, model)
    383 
    384         # add samplers
--> 385         dataloaders = [self._prepare_dataloader(dl, mode=mode) for dl in dataloaders if dl is not None]
    386 
    387         # add worker_init_fn for correct seeding in worker processes

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/connectors/data_connector.py in <listcomp>(.0)
    383 
    384         # add samplers
--> 385         dataloaders = [self._prepare_dataloader(dl, mode=mode) for dl in dataloaders if dl is not None]
    386 
    387         # add worker_init_fn for correct seeding in worker processes

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/trainer/connectors/data_connector.py in _prepare_dataloader(self, dataloader, shuffle, mode)
    294 
    295             sampler = self._resolve_sampler(dataloader, shuffle=shuffle, mode=mode)
--> 296             dataloader = _update_dataloader(dataloader, sampler, mode=mode)
    297 
    298         dataloader = self.trainer.strategy.process_dataloader(dataloader)

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/utilities/data.py in _update_dataloader(dataloader, sampler, mode)
    200     dataloader: DataLoader, sampler: Union[Sampler, Iterable], mode: Optional[RunningStage] = None
    201 ) -> DataLoader:
--> 202     dl_args, dl_kwargs = _get_dataloader_init_args_and_kwargs(dataloader, sampler, mode)
    203     dataloader = _reinstantiate_wrapped_cls(dataloader, *dl_args, **dl_kwargs)
    204     return dataloader

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/utilities/data.py in _get_dataloader_init_args_and_kwargs(dataloader, sampler, mode, disallow_batch_sampler)
    248     if not was_wrapped:
    249         # keep only the params whose default is different to the current attr value
--> 250         non_defaults = {name for name, p in params.items() if name in attrs and p.default != attrs[name]}
    251 
    252         # add `dataset` as it might have been replaced with `*args`

~/.miniconda/lib/python3.7/site-packages/pytorch_lightning/utilities/data.py in <setcomp>(.0)
    248     if not was_wrapped:
    249         # keep only the params whose default is different to the current attr value
--> 250         non_defaults = {name for name, p in params.items() if name in attrs and p.default != attrs[name]}
    251 
    252         # add `dataset` as it might have been replaced with `*args`

ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()



```

### Environment

```

#- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow): Trainer
#- PyTorch Lightning Version (e.g., 1.5.0): 1.7.7
#- Lightning App Version (e.g., 0.5.2):
#- PyTorch Version (e.g., 1.10): 1.12.1.post200
#- Python version (e.g., 3.9): 3.7.12
#- OS (e.g., Linux): Linux
#- CUDA/cuDNN version:
#- GPU models and configuration:
#- How you installed Lightning(`conda`, `pip`, source):
#- Running environment of LightningApp (e.g. local, cloud):

```

cc @justusschock @awaelchli @ninginthecloud @rohitgr7 @otaj

**Comment by gokceneraslan** [comment]
Thanks!

**Comment by awaelchli** [comment]
@gokceneraslan Thanks for reporting this. The fix for this will be released in the Lightning 1.8 package very shortly.
